File: dining-savages.py
from threading import Thread, Semaphore, Lock

savages = 12
M = 10
servings = 0
filled = Semaphore(0)
empty = Semaphore(0)
lock = Lock()

class Cook(Thread):
    def run(self):
        while True:
            empty.acquire()
            print("Filling pot")
            filled.release()


class Savage(Thread):
    def __init__(self, i):
        self.i = i
        super(Savage, self).__init__()
        print("Savage " + str(i) + "made")
    
    def run(self):
        global servings
        while True:
            lock.acquire()
            if servings == 0:
                empty.release()
                filled.acquire()
                servings = M
            servings -= 1
            print("Savage " + str(i) + " has ate")
            lock.release()


cook = Cook()
cook.start()
for i in range(savages):
    sav = Savage(i)
    sav.start()

cook.join()
    

File: dining-savages.go
package main

import (
	"fmt"
	"sync"
)

var M = 10
var num_savages = 12
var servings = 0

type Semaphore chan struct{}

func (sem Semaphore) Wait() {
	<-sem
}

func (sem Semaphore) Signal() {
	sem <- struct{}{}
}

var empty = make(Semaphore)
var full = make(Semaphore)
var mutex = sync.Mutex{}

func cook() {
	for true {
		empty.Wait()
		fmt.Println("Filling pot")
		full.Signal()
	}
}

func savage(i int) {
	for true {
		mutex.Lock()
		if servings == 0 {
			empty.Signal()
			full.Wait()
			servings = M
		}
		servings -= 1
		fmt.Printf("Savage %d ate from the pot\n", i)
		mutex.Unlock()
	}

}

func main() {
	go cook()
	for i := 0; i < num_savages; i++ {
		go savage(i)
	}
	select {}
}
